File: aho_corasick.py
import time
from collections import deque

# Node class

class Node:
    """
    Reprezentuje węzeł automatu Aho–Corasick.

    Węzeł przechowuje przejścia znakowe, łącze porażki (failure link),
    listę identyfikatorów wzorców kończących się w tym węźle oraz
    identyfikator numeryczny używany przy wizualizacji grafu.
    """

    __slots__ = ("next", "fail", "out", "id")

    def __init__(self):
        """
        Inicjalizuje pusty węzeł automatu.
        """
        self.next = {}
        self.fail = None
        self.out = []
        self.id = None


# Aho–Corasick automaton

class AhoCorasick:
    """
    Implementacja automatu Aho–Corasick do wielowzorcowego
    wyszukiwania podciągów w tekście.

    Automat umożliwia dodawanie wzorców, budowę struktury
    przejść oraz efektywne przeszukiwanie tekstu w czasie liniowym.
    """

    def __init__(self):
        """
        Tworzy pusty automat z węzłem korzenia.
        """
        self.root = Node()
        self._nodes = [self.root]

    def add(self, pattern, pat_id=None):
        """
        Dodaje wzorzec do struktury trie automatu.

        :param pattern: Wzorzec znakowy do dodania.
        :type pattern: str
        :param pat_id: Identyfikator wzorca zapisywany w liście wyjść.
                       Jeżeli None, zapisywany jest sam wzorzec.
        :type pat_id: object | None
        :return: None
        :rtype: None
        """
        node = self.root
        for ch in pattern:
            if ch not in node.next:
                new = Node()
                node.next[ch] = new
                self._nodes.append(new)
            node = node.next[ch]
        node.out.append(pat_id if pat_id is not None else pattern)

    def build(self):
        """
        Buduje automat Aho–Corasick na podstawie dodanych wzorców.

        Operacja obejmuje:
        - wyznaczenie łączy porażki (failure links),
        - propagację list wyjść,
        - przypisanie identyfikatorów numerycznych węzłom.

        :return: None
        :rtype: None
        """
        queue = deque()
        self.root.fail = self.root

        for ch, nxt in self.root.next.items():
            nxt.fail = self.root
            queue.append(nxt)

        while queue:
            r = queue.popleft()

            for ch, u in r.next.items():
                queue.append(u)
                v = r.fail

                while v is not self.root and ch not in v.next:
                    v = v.fail

                u.fail = v.next[ch] if ch in v.next else self.root
                u.out += u.fail.out

        for i, n in enumerate(self._nodes):
            n.id = i

    def search(self, text, yield_matches=False):
        """
        Przeszukuje tekst przy użyciu zbudowanego automatu.

        Znaki spoza alfabetu ``ACGTN`` powodują powrót do korzenia
        automatu i są pomijane w dopasowaniach.

        :param text: Tekst wejściowy do przeszukania.
        :type text: str
        :param yield_matches: Jeżeli True, zwracana jest również lista
                              szczegółowych dopasowań.
        :type yield_matches: bool
        :return: Liczba dopasowań lub krotka (liczba, lista_dopasowań).
        :rtype: int | tuple[int, list[tuple[int, object]]]
        """
        node = self.root
        total = 0
        matches = [] if yield_matches else None
        idx = 0

        for ch in text:
            if ch not in "ACGTN":
                node = self.root
                idx += 1
                continue

            while node is not self.root and ch not in node.next:
                node = node.fail

            node = node.next[ch] if ch in node.next else self.root

            if node.out:
                total += len(node.out)
                if yield_matches:
                    for o in node.out:
                        matches.append((idx, o))

            idx += 1

        return (total, matches) if yield_matches else total


def load_fasta(path):
    """
    Wczytuje sekwencję z pliku FASTA i zwraca ją jako pojedynczy łańcuch.

    Linie nagłówkowe (rozpoczynające się od znaku ``>``) są pomijane.

    :param path: Ścieżka do pliku FASTA.
    :type path: str
    :return: Połączona sekwencja w wielkich literach.
    :rtype: str
    """
    seqs = []
    with open(path, "r") as f:
        buf = []
        for line in f:
            if line.startswith(">"):
                if buf:
                    seqs.append("".join(buf).upper())
                    buf = []
            else:
                buf.append(line.strip())
        if buf:
            seqs.append("".join(buf).upper())
    return "".join(seqs)


def load_patterns(path):
    """
    Wczytuje wzorce z pliku tekstowego.

    Pomijane są linie puste oraz zawierające znaki spoza alfabetu ``ACGTN``.

    :param path: Ścieżka do pliku ze wzorcami.
    :type path: str
    :return: Lista poprawnych wzorców.
    :rtype: list[str]
    """
    pats = []
    with open(path, "r") as f:
        for line in f:
            s = line.strip().upper()
            if not s:
                continue
            if any(ch not in "ACGTN" for ch in s):
                continue
            pats.append(s)
    return pats


def export_dot(ac, filename="automaton.dot", limit_nodes=1500):
    """
    Eksportuje strukturę automatu do pliku DOT (Graphviz).

    :param ac: Zbudowany automat Aho–Corasick.
    :type ac: AhoCorasick
    :param filename: Nazwa pliku wyjściowego.
    :type filename: str
    :param limit_nodes: Maksymalna liczba węzłów do zapisania.
    :type limit_nodes: int
    :return: None
    :rtype: None
    """
    nodes = ac._nodes
    N = min(len(nodes), limit_nodes)

    with open(filename, "w") as f:
        f.write("digraph aho {\n")
        f.write("  rankdir=LR;\n")
        f.write("  node [shape=circle,fontname=Helvetica];\n")

        for i in range(N):
            if nodes[i].out:
                f.write(f'  n{i} [label="{i}\\nout={len(nodes[i].out)}",style=filled,fillcolor=lightblue];\n')
            else:
                f.write(f'  n{i} [label="{i}"];\n')

        for i in range(N):
            for ch, nxt in nodes[i].next.items():
                if nxt.id < N:
                    f.write(f'  n{i} -> n{nxt.id} [label="{ch}"];\n')

        for i in range(N):
            fail = nodes[i].fail
            if fail and fail.id < N and nodes[i] is not ac.root:
                f.write(f'  n{i} -> n{fail.id} [style=dashed,color=gray,label="f"];\n')

        f.write("}\n")

    print(f"[OK] DOT automaton saved to {filename}")


def benchmark(fasta_path, patterns_path):
    """
    Wykonuje pełny benchmark działania automatu Aho–Corasick.

    :param fasta_path: Ścieżka do pliku FASTA.
    :type fasta_path: str
    :param patterns_path: Ścieżka do pliku ze wzorcami.
    :type patterns_path: str
    :return: None
    :rtype: None
    """
    print("=== Benchmark: Aho–Corasick ===")
    print("Loading FASTA...")
    text = load_fasta(fasta_path)
    print(f"Text length: {len(text)}")

    print("Loading patterns...")
    pats = load_patterns(patterns_path)
    print(f"Patterns: {len(pats)}")

    ac = AhoCorasick()

    print("Adding patterns...")
    for i, p in enumerate(pats):
        ac.add(p, i)

    print("Building automaton...")
    t0 = time.perf_counter()
    ac.build()
    t1 = time.perf_counter()
    print(f"Build time: {t1 - t0:.3f} s; nodes = {len(ac._nodes)}")

    print("Searching...")
    t0 = time.perf_counter()
    matches = ac.search(text)
    t1 = time.perf_counter()
    print(f"Search time: {t1 - t0:.3f} s")
    print(f"Total matches: {matches}")

    export_dot(ac, "automaton.dot")

File: aho_corasick.cpp
/**
 * @file aho_corasick.cpp
 * @author Maria Ławniczak (Nr Indeksu: 268544)
 * @brief Implementacja algorytmu Aho-Corasick dla sekwencji DNA
 * @date 2026-01-25
 */

#include <bits/stdc++.h>

using namespace std;

/**
 * @brief Mapuje znaki alfabetu DNA na indeksy tablicy (0-4)
 * Obsługuje A, C, G, T oraz N (jako błąd lub nieznany nukleotyd)
 * * @param c Znak do zmapowania
 * @return int Indeks z zakresu 0-4
 */
int char_idx(char c) {
    switch (toupper(c)) {
        case 'A': return 0;
        case 'C': return 1;
        case 'G': return 2;
        case 'T': return 3;
        case 'N': return 4;
        default:  return 4; 
    }
}

/**
 * @brief Wczytuje wzorce z pliku tekstowego, czyszcząc je z białych znaków
 * * @param path Ścieżka do pliku
 * @return vector<string> Lista oczyszczonych wzorców
 */
vector<string> load_patterns(const string& path) {
    ifstream in(path);
    if (!in) {
        cerr << "Blad: Nie mozna otworzyc pliku " << path << "\n";
        exit(1);
    }
    string line;
    vector<string> out;
    while (getline(in, line)) {
        string clean;
        for (char c : line) {
            if (!isspace((unsigned char)c))
                clean.push_back(toupper(c));
        }
        if (!clean.empty()) out.push_back(clean);
    }
    return out;
}

/**
 * @brief Struktura reprezentująca pojedynczy węzeł w automacie Aho-Corasick
 */
struct Node {
    array<int, 5> next;  // Przejścia do kolejnych stanów dla alfabetu {A, C, G, T, N}
    int fail;            // Wskaźnik funkcji porażki
    vector<int> out;     ///Lista identyfikatorów wzorców, które kończą się w tym stanie

    Node() {
        next.fill(-1);
        fail = 0;
    }
};

int main(int argc, char** argv) {
    if (argc < 2) {
        cerr << "Uzycie: " << argv[0] << " <wzorce.txt> [eksport.dot]\n";
        return 1;
    }

    // Inicjalizacja danych
    vector<string> pats = load_patterns(argv[1]);
    vector<Node> trie(1); // Stan 0 to root

    // Budowa drzewa Trie
    for (size_t pid = 0; pid < pats.size(); pid++) {
        int v = 0;
        for (char c : pats[pid]) {
            int id = char_idx(c);
            if (trie[v].next[id] == -1) {
                trie[v].next[id] = trie.size();
                trie.emplace_back();
            }
            v = trie[v].next[id];
        }
        trie[v].out.push_back(pid); // Oznaczamy koniec wzorca w danym węźle
    }

    // Budowa funkcji porażki
    queue<int> q;

    // Inicjalizacja poziomu 1 (bezpośredni sąsiedzi roota)
    for (int c = 0; c < 5; c++) {
        int nxt = trie[0].next[c];
        if (nxt != -1 && nxt != 0) {
            trie[nxt].fail = 0;
            q.push(nxt);
        } else {
            trie[0].next[c] = 0; // Jeśli brak przejścia z roota, wracamy do roota
        }
    }

    // Przetwarzanie kolejnych poziomów drzewa
    while (!q.empty()) {
        int r = q.front();
        q.pop();

        for (int c = 0; c < 5; c++) {
            int u = trie[r].next[c];
            if (u == -1 || u == 0) continue;

            q.push(u);
            int f = trie[r].fail;

            // Szukamy najdłuższego właściwego sufiksu, który jest w Trie
            while (trie[f].next[c] == -1) {
                f = trie[f].fail;
            }

            trie[u].fail = trie[f].next[c];
    
            // Jeśli węzeł, do którego prowadzi fail-link, jest końcem wzorca, to obecny węzeł również "zawiera" ten wzorzec
            for (int id : trie[trie[u].fail].out) {
                trie[u].out.push_back(id);
            }
        }
    }

    cerr << "Statystyki automatu:\n";
    cerr << " - Liczba wzorcow: " << pats.size() << "\n";
    cerr << " - Liczba wezlow (stanow): " << trie.size() << "\n";

    // Eksport do formatu DOT 
    if (argc >= 3) {
        ofstream f(argv[2]);
        if (!f) {
            cerr << "Nie mozna utworzyc pliku .dot!\n";
            return 1;
        }

        f << "digraph AC {\n"
          << "  rankdir=LR;\n"
          << "  node [shape=circle];\n";

        for (size_t i = 0; i < trie.size(); i++) {
            if (trie[i].out.empty()) {
                f << "  n" << i << " [label=\"" << i << "\"];\n";
            } else {
                f << "  n" << i << " [label=\"" << i
                  << "\\n(Pats: " << trie[i].out.size()
                  << ")\", style=filled, fillcolor=lightblue];\n";
            }
        }

        // Krawędzie drzewa (przejścia)
        for (size_t i = 0; i < trie.size(); i++) {
            for (int c = 0; c < 5; c++) {
                int v = trie[i].next[c];
                if (v != -1 && v != 0) {
                    f << "  n" << i << " -> n" << v << " [label=\"" << "ACGTN"[c] << "\"];\n";
                }
            }
        }

        // Krawędzie fail-linków (przerywane)
        for (size_t i = 1; i < trie.size(); i++) {
            f << "  n" << i << " -> n" << trie[i].fail
              << " [style=dashed, color=red, label=\"fail\"];\n";
        }

        f << "}\n";
        cerr << "Zapisano graf do: " << argv[2] << "\n";
    }

    return 0;
}
